docs/tutorials/hpo-workflows/hpo-workflows.rst

```diff
@@ -0,0 +1,334 @@
+                    )
+                    # First, define strategy-wise optional configurations
+                    if isinstance(self.strategy, DeepSpeedStrategy):
+                        distribute_kwargs = dict(
+                            config_params=dict(train_micro_batch_size_per_gpu=self.config.batch_size)
+                        )
+                    else:
+                        distribute_kwargs = {}
+                    optimizer = Adam(model.parameters(), lr=self.config.learning_rate)
+                    self.model, self.optimizer, _ = self.strategy.distributed(
+                        model, optimizer, **distribute_kwargs
+                    )
+                    self.loss = CrossEntropyLoss()
+
+                def train(self, config, data):
+                    
+                    for epoch in range(self.config.epochs):
+                        if self.strategy.global_world_size() > 1:
+                            self.set_epoch(epoch)
+
+                        train_losses = []
+                        val_losses = []
+
+                        for images, labels in enumerate(self.train_dataloader):
+                            
+                            device = self.strategy.device()
+                            images, labels = images.to(device), labels.to(device)
+
+                            outputs = self.model(images)
+                            train_loss = self.loss(outputs, labels)
+                            self.optimizer.zero_grad()
+                            train_loss.backward()
+                            self.optimizer.step()
+                            train_losses.append(train_loss.detach().cpu().numpy())
+
+                        for images, labels in enumerate(self.validation_dataloader):
+
+                            device = self.strategy.device()
+                            images, labels = images.to(device), labels.to(device)
+
+                            with torch.no_grad():
+                                outputs = self.model(images)
+                                val_loss = self.loss(outputs, labels)
+                            val_losses.append(val_loss.detach().cpu().numpy())
+
+                        self.log(np.mean(train_losses), "train_loss", kind="metric", step=epoch)
+                        self.log(np.mean(val_losses), "val_loss", kind="metric", step=epoch)
+                        checkpoint = {
+                            "epoch": epoch,
+                            "loss": train_loss,
+                            "val_loss": val_loss,
+                        }
+                        checkpoint_filename = self.checkpoints_location.format(epoch)
+                        torch.save(checkpoint, checkpoint_filename)
+                        self.log(
+                            checkpoint_filename,
+                            os.path.basename(checkpoint_filename),
+                            kind="artifact",
+                        )
```

Comment:
This is great! However, for the future, please consider that embedding code in the docs requires that changes in the code base are reflected in the docs. But this is fine

Reply:
Okay will do!